src/rules/contrast.ts

```diff
@@ -0,0 +1,100 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { isElementVisible } from "../utils";
+import { ValidationRule, ValidationResult, ValidationRuleType } from "./base";
+
+function hexToRgb(hex: string): [number, number, number] | null {
+  hex = hex.replace(/^#/, "");
+  if (hex.length === 3) {
+    hex = hex
+      .split("")
+      .map((x) => x + x)
+      .join("");
+  }
+  if (hex.length !== 6) {
+    return null;
+  }
+  const num = parseInt(hex, 16);
+  return [(num >> 16) & 255, (num >> 8) & 255, num & 255];
+}
+
+function parseColor(color: string): [number, number, number] | null {
+  color = color.trim();
+  if (color.startsWith("#")) {
+    return hexToRgb(color);
+  }
+  const rgbMatch = color.match(/^rgb\((\d+),\s*(\d+),\s*(\d+)\)$/);
+  if (rgbMatch) {
+    return [
+      parseInt(rgbMatch[1], 10),
+      parseInt(rgbMatch[2], 10),
+      parseInt(rgbMatch[3], 10),
+    ];
+  }
+  return null;
+}
+
+function luminance([r, g, b]: [number, number, number]): number {
+  const a = [r, g, b].map((v) => {
+    v /= 255;
+    return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
+  });
+  return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
+}
+
+function contrastRatio(l1: number, l2: number): number {
+  return (Math.max(l1, l2) + 0.05) / (Math.min(l1, l2) + 0.05);
+}
+
+export class ContrastRule extends ValidationRule {
+  type = ValidationRuleType.Error;
+  name = "ContrastRule";
+  anchored = true;
+
+  accept(element: HTMLElement): boolean {
+    return isElementVisible(element) && !!element.textContent?.trim();
+  }
+
+  validate(element: HTMLElement): ValidationResult | null {
+    const style = window.getComputedStyle(element);
```

Comment:
is this always the right `window`?
what is the perf of getComputedStyle - do we need to think about delaying the computation somehow?

Reply:
About the window... I think so, it's similar to what other rules are doing like `focusLost`, I can totally do some validation if the window does exist also like in the `focusLost` .

The `getComputedStyle` is also something we are already using 

https://github.com/microsoft/abledom/blob/5e5b7d43c29a673ad5436b9bdd91731d9dcf559e/src/utils.ts#L187

But I would be curious if @mshoho Has some ideas how to improve this here